Add release promotion endpoint in Android-Components

sarah-clements opened the pull request:
This is a WIP for [this ](https://jira.mozilla.com/browse/RELENG-149) sprint item. @JohanLorenzo I have a question (see comment below) and will continue with tests and reading about hooks and CI configuration per your suggestion.

sarah-clements commented on taskcluster/ac_taskgraph/release_promotion.py:
```diff
@@ -0,0 +1,172 @@
+
+import os
+
+from mozilla_version.fenix import FenixVersion
```
What should this be?

JohanLorenzo replied:
Great question! It should be:

```suggestion
 from mozilla_version.maven import MavenVersion 
```

I concede it's not obvious to guess 😅 Here's the context: Fenix is an end-product which used to follow its own version scheme. Now it's almost on track with Gecko versions. Android-components are Android libraries, that is to say: Java libraries packaged by the tool called maven. Maven defines a strict way of defining version numbers which don't match neither Fenix nor Gecko. That's why we should use `MavenVersion`

sarah-clements replied:
Gotcha, thanks for explaining that detail!

JohanLorenzo reviewed (commented):
Sounds good! Let me know if you'd like some pointers/more context 🙂

escapewindow commented on taskcluster/ac_taskgraph/release_promotion.py:
```diff
@@ -0,0 +1,172 @@
+                'type': 'string',
+                'description': 'The flavor of release promotion to perform.',
+                'default': 'build',
+                'enum': sorted(graph_config['release-promotion']['flavors'].keys()),
```
Ah. To define `graph_config['release-promotion']['flavors']`, we probably need to copy [this block](https://github.com/mozilla-mobile/fenix/blob/master/taskcluster/ci/config.yml#L96-L99) into `taskcluster/ci/config.yml` in android-components.

sarah-clements replied:
Thanks for the pointer!

escapewindow commented on taskcluster/ac_taskgraph/__init__.py:
```diff
@@ -23,7 +23,7 @@ def register(graph_config):
     the process.
     """
-    _import_modules(["job", "routes", "target_tasks", "worker_types"])
+    _import_modules(["job", "routes", "target_tasks", "worker_types", "release-promotion"])
```
I think you need an underscore here. `release_promotion`. This is probably what's causing the current decision task bustage, `[task 2020-12-04T00:56:51.079Z] ImportError: No module named release-promotion`

sarah-clements replied:
oh right, my bad!

escapewindow reviewed (commented):
Just the two comments + the `s,FenixVersion,MavenVersion` and I think we'll be code complete. We may have to add some scopes to get this to work properly, but I think we can find those during our staging tests. Thank you!

escapewindow commented on taskcluster/ac_taskgraph/release_promotion.py:
```diff
@@ -0,0 +1,172 @@
+
+    parameters['next_version'] = input['next_version']
+
+    version = FenixVersion.parse(version_string)
```
We probably also need to change this `FenixVersion` into `MavenVersion`.

sarah-clements replied:
hah, right. fixed.

escapewindow reviewed (commented):
So we're dying on importing `MavenVersion`:

```
[task 2020-12-04T21:42:49.582Z]     from mozilla_version.maven import MavenVersion
[task 2020-12-04T21:42:49.583Z] ImportError: No module named mozilla_version.maven
```

This probably means we need to install the `mozilla-version` dependency before we run the decision task. I'm not seeing where we do this in Fenix. This may be a question for @JohanLorenzo :)

sarah-clements commented:
> Sounds good! Let me know if you'd like some pointers/more context 🙂

I didn't actually get to the tests today. If you could give me pointers on where to write them and what you had in mind (testing that this action works??) that'd be great. I did some poking around in Fenix to see what tests for this would've looked like but it's not obvious to me and this repo doesn't seem to have any documentation on writing tests.

escapewindow commented:
Johan might have a different answer, but I don't think we have any automated unit tests if that's what you're talking about. I would LOVE to see 100% unit test coverage in taskgraph, but we haven't prioritized that.

Most likely we'll be testing manually. One way is to use `taskgraph test-action-callback` locally... this is super useful, but not necessarily intuitive. Likely we'll need to cross-train here. Another is to test against a repo, in taskcluster. I think that once I resolve the scopes issues in my fork, I can land your patches there, and we can try a staging release from shipit-staging using cam's patches. I'd like to set up a shared staging repo at some point, so we can each test our patches without relying on others. I'll add that as a lower priority item in the epic.

JohanLorenzo commented on taskcluster/ac_taskgraph/release_promotion.py:
```diff
@@ -0,0 +1,172 @@
+
+import os
+
+from mozilla_version.maven import MavenVersion
```
Bug:

> This probably means we need to install the mozilla-version dependency before we run the decision task. I'm not seeing where we do this in Fenix. This may be a question for @JohanLorenzo :)

Ah yeah, this something we hit on Fenix. We had to install the `mozilla-version` package in the decision task at these 2 places: 

* https://github.com/mozilla-mobile/fenix/blob/2cada405d81fe4831e866794c91e57fb088315ac/.taskcluster.yml#L264
* https://github.com/mozilla-mobile/fenix/blob/2cada405d81fe4831e866794c91e57fb088315ac/.taskcluster.yml#L270

We also have to bump the docker image defined in `.taskcluster.yml`: 

https://github.com/mozilla-mobile/fenix/commit/7d3b23ceeb12c0e3edcb36836e9041b9ca7a1800#diff-a728f7e52d751b98eafa856e45594533339b44f229d7b83f930df335391e7e15R246

The reasons is not super explicit, but let me explain:

 * The decision task uses a docker image that already contains some python packages, and notably `attrs`
 * `mozilla-version` requires an up-to-date version of `attrs`
 * on the current docker image, attrs is too old and does not get updated. See line 64 of `live_backing.log` https://firefox-ci-tc.services.mozilla.com/tasks/fM6I_bXkQeeANWIBKquWsA
 * the new docker image has a compatible `attrs`


@escapewindow, in the long term, we likely want to bake this package in https://hg.mozilla.org/ci/taskgraph/file/2b2622598df02bde211d8cedb334b7b22fb883a4/taskcluster/docker/decision-mobile/Dockerfile

sarah-clements replied:
ok, I've done this.

JohanLorenzo commented on taskcluster/ac_taskgraph/release_promotion.py:
```diff
@@ -0,0 +1,172 @@
+
+
+def read_version_file():
+    with open(os.path.join(os.path.dirname(__file__), '..', '..', 'version.txt')) as f:
```
Bug: This file doesn't exist yet. We'll have to create it and to remove this line:

https://github.com/mozilla-mobile/android-components/blob/35c99f4ba90532ec301df4148ac49be48b902004/.buildconfig.yml#L1

Some things are depending on `.buildconfig.yml` so we'll have to update calls to use `version.txt` instead. @escapewindow asked whether we should use `.buildconfig.yml` or `version.txt`. I believe the latter is best. For more context: see https://jira.mozilla.com/browse/RELENG-152?focusedCommentId=197982&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-197982

sarah-clements replied:
So, if I'm following this correctly is the issue [146](https://jira.mozilla.com/browse/RELENG-146) that adds `version.txt` a blocker to landing this??

JohanLorenzo replied:
I'm sorry, I may not follow what you mean regarding [146](https://jira.mozilla.com/browse/RELENG-146). My apologies if my answer doesn't address your point.

I see 2 ways of solving this issue:
 1. either we add `version.txt` and stop using the value in `.buildconfig.yml` 
 2. or we update `read_version_file()` to use `.buildconfig.yml` now , then we do point 1 in a follow up.

I don't have any preference.

sarah-clements replied:
Aki said he will be adding `version.text` in a separate pr so we'll coordinate the merging of his first.

JohanLorenzo replied:
I'm sorry I missed your last message.

Awesome! I see Aki already landed the patch, so we're all good here! Thank you for the context 🙂

JohanLorenzo commented on taskcluster/ac_taskgraph/release_promotion.py:
```diff
@@ -0,0 +1,172 @@
+
+    version = MavenVersion.parse(version_string)
+    if version.is_beta:
+        release_type = "beta"
```
Nit: Android-components don't support `beta`. There's only `nightly` and `release`. We can remove this path.

sarah-clements replied:
done

JohanLorenzo commented on taskcluster/ac_taskgraph/release_promotion.py:
```diff
@@ -0,0 +1,172 @@
+from taskgraph.util.taskgraph import find_decision_task, find_existing_tasks_from_previous_kinds
+
+RELEASE_PROMOTION_PROJECTS = (
+    "https://github.com/mozilla-mobile/android-components",
```
Bug: We have to add `"https://github.com/escapewindow/android-components"` for the initial end-to-end testing. Otherwise `actions.json` won't contain the `release-promotion` action. See https://firefox-ci-tc.services.mozilla.com/tasks/YVymMAawR56yGmBB35ZerQ#artifacts, for instance

JohanLorenzo reviewed (commented):
> I resolve the scopes issues in my fork, I can land your patches there, and we can try a staging release from shipit-staging using Cam's patches. 

I went that route. It enabled me to test the integration between this PR and https://github.com/mozilla-releng/shipit/pull/547. We're on the right way. I'm currently seeing this stacktrace on shipit. I believe it's because of `"https://github.com/escapewindow/android-components"` is missing from `RELEASE_PROMOTION_PROJECTS`. Once that's fixed, we should be closer to get a release promotion started.

If you would like to test this locally too: 

1. Clone shipit
2. follow https://github.com/mozilla-releng/shipit/#local-development
3. Checkout @camd's patch. 
4. Go to https://localhost:8010/new and select: 
 * Product: Android-Components
 * Repositories: Aki's fork
 * Branch: master
 * Revision: [whatever is on top of that branch]
5. Then click on `Create release` 

You will likely get an error for now, have a look at your terminal if that's the case.

```
api_1     | 2020-12-07 18:27:13,989 - INFO: Looking up taskcluster index: garbage.mobile.v2.android-components.revision.f431181621898da999f22c3ab216f6038e9b6601.taskgraph.decision
api_1     | 2020-12-07 18:27:16,083 - ERROR: 'NoneType' object is not subscriptable
api_1     | 2020-12-07 18:27:16,084 - ERROR: Exception on /releases [POST]
api_1     | Traceback (most recent call last):
api_1     |   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
api_1     |     response = self.full_dispatch_request()
api_1     |   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
api_1     |     rv = self.handle_user_exception(e)
api_1     |   File "/usr/local/lib/python3.8/site-packages/flask_cors/extension.py", line 165, in wrapped_function
api_1     |     return cors_after_request(app.make_response(f(*args, **kwargs)))
api_1     |   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
api_1     |     reraise(exc_type, exc_value, tb)
api_1     |   File "/usr/local/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
api_1     |     raise value
api_1     |   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
api_1     |     rv = self.dispatch_request()
api_1     |   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
api_1     |     return self.view_functions[rule.endpoint](**req.view_args)
api_1     |   File "/usr/local/lib/python3.8/site-packages/connexion/decorators/decorator.py", line 48, in wrapper
api_1     |     response = function(request)
api_1     |   File "/usr/local/lib/python3.8/site-packages/connexion/decorators/uri_parsing.py", line 144, in wrapper
api_1     |     response = function(request)
api_1     |   File "/usr/local/lib/python3.8/site-packages/connexion/decorators/validation.py", line 184, in wrapper
api_1     |     response = function(request)
api_1     |   File "/usr/local/lib/python3.8/site-packages/connexion/decorators/response.py", line 103, in wrapper
api_1     |     response = function(request)
api_1     |   File "/usr/local/lib/python3.8/site-packages/connexion/decorators/parameter.py", line 121, in wrapper
api_1     |     return function(**kwargs)
api_1     |   File "/usr/local/lib/python3.8/site-packages/shipit_api/admin/api.py", line 110, in add_release
api_1     |     release.phases = generate_phases(release, common_input, verify_supported_flavors=True)
api_1     |   File "/usr/local/lib/python3.8/site-packages/shipit_api/admin/tasks.py", line 145, in generate_phases
api_1     |     for phase in release_promotion_flavors(release, actions, verify_supported_flavors):
api_1     |   File "/usr/local/lib/python3.8/site-packages/shipit_api/admin/tasks.py", line 175, in release_promotion_flavors
api_1     |     avail_flavors = relpro["schema"]["properties"]["release_promotion_flavor"]["enum"]
api_1     | TypeError: 'NoneType' object is not subscriptable
```

sarah-clements commented:
> > I resolve the scopes issues in my fork, I can land your patches there, and we can try a staging release from shipit-staging using Cam's patches.
> 
> I went that route. It enabled me to test the integration between this PR and [mozilla-releng/shipit#547](https://github.com/mozilla-releng/shipit/pull/547). We're on the right way. I'm currently seeing this stacktrace on shipit. I believe it's because of `"https://github.com/escapewindow/android-components"` is missing from `RELEASE_PROMOTION_PROJECTS`. Once that's fixed, we should be closer to get a release promotion started.
> 
> If you would like to test this locally too:
> 
>     1. Clone shipit
> 
>     2. follow https://github.com/mozilla-releng/shipit/#local-development
> 
>     3. Checkout @camd's patch.
> 
>     4. Go to https://localhost:8010/new and select:
> 
> 
>     * Product: Android-Components
> 
>     * Repositories: Aki's fork
> 
>     * Branch: master
> 
>     * Revision: [whatever is on top of that branch]
> 
> 
>     1. Then click on `Create release`
> 
So, I'm missing some critical understanding here. Firstly, how does this actually test this pr/branch? Does it need to be pushed to Aki's fork?

Secondly, I got to Step 4 but then after seeing the generic VPN message and checking the network tab (only saw 200 response codes), I couldn't actually click on anything. Where do I select the product? I tried signing in but got an error there too.

![Screen Shot 2020-12-07 at 4 50 56 PM](https://user-images.githubusercontent.com/19615783/101422925-b8bf3080-38ac-11eb-9348-e66005070c15.png)

JohanLorenzo commented:
Thank you for your questions! 

> So, I'm missing some critical understanding here. Firstly, how does this actually test this pr/branch? Does it need to be pushed to Aki's fork?

Yeah, Aki's fork will be easier to test this out. There is another way, but it'll require some knowledge about Taskcluster and hooks + some mental tweaks.

> Secondly, I got to Step 4 but then after seeing the generic VPN message and checking the network tab (only saw 200 response codes), I couldn't actually click on anything. Where do I select the product? I tried signing in but got an error there too.

🤔 The VPN could be caused by the CORS requests that don't get through. Although, if you see 200s, then you should be all set. I know for sure you need to be logged in. What error do you get on the docker-compose terminal when you do so? 
I know @camd struggled to get things working locally, so there may be something we missed. I know he's pushing changes to the staging instance, so that may unblock you with testing, in the short term.

If that still doesn't work, we can have a call tomorrow and debug things together.

sarah-clements commented:
> Thank you for your questions!
> 
> > So, I'm missing some critical understanding here. Firstly, how does this actually test this pr/branch? Does it need to be pushed to Aki's fork?
> 
> Yeah, Aki's fork will be easier to test this out. There is another way, but it'll require some knowledge about Taskcluster and hooks + some mental tweaks.
> 
> > Secondly, I got to Step 4 but then after seeing the generic VPN message and checking the network tab (only saw 200 response codes), I couldn't actually click on anything. Where do I select the product? I tried signing in but got an error there too.
> 
> 🤔 The VPN could be caused by the CORS requests that don't get through. Although, if you see 200s, then you should be all set. I know for sure you need to be logged in. What error do you get on the docker-compose terminal when you do so?
> I know @camd struggled to get things working locally, so there may be something we missed. I know he's pushing changes to the staging instance, so that may unblock you with testing, in the short term.
> 
> If that still doesn't work, we can have a call tomorrow and debug things together.

I didn't get a chance to try this again but at least in the meantime I pushed my changes to Aki's fork for testing.

escapewindow commented:
I landed https://github.com/escapewindow/android-components/commit/ba0dd6e8be6e2b8d6e6484c97ec322145cc95d9c and https://github.com/escapewindow/android-components/commit/6f1694bd1b034be54c486c24c24d098e42008df3 in my fork to fix a couple of things here. That may or may not be everything -- we'll know more when we're green end-to-end.

escapewindow commented:
I merged #9077, which should resolve the version.txt issue.

escapewindow commented:
Ok, https://github.com/mozilla-mobile/android-components/pull/9202 has landed with `version.txt` :)

sarah-clements commented:
> Ok, #9202 has landed with `version.txt` :)

I've rebased and added the changes mentioned in https://github.com/mozilla-mobile/android-components/pull/9132#issuecomment-741492448

escapewindow reviewed (approved):
Thank you! I also added https://github.com/escapewindow/android-components/commit/13211fe8ad0d0db3977d54d5aa750275b431ac50#diff-3b7ed13f881c4a6e84fbfa47ce84c8c7930501acfd76fcef8d0f11a4799598a6 in staging, but that doesn't block.

This looks good to me. I may leave this for @JohanLorenzo to take a look as well. I have a todo item to demo how to debug an in-tree action.

JohanLorenzo commented on taskcluster/ac_taskgraph/release_promotion.py:
```diff
@@ -0,0 +1,163 @@
+from taskgraph.util.taskgraph import find_decision_task, find_existing_tasks_from_previous_kinds
+
+RELEASE_PROMOTION_PROJECTS = (
+    "https://github.com/mozilla-mobile/android-components", "https://github.com/escapewindow/android-components"
```
Note to our future selves: Remove `https://github.com/escapewindow/android-components` once we're done with staging releases.

JohanLorenzo reviewed (approved):
This looks great to me! We tested this on stating in https://firefox-ci-tc.services.mozilla.com/tasks/KzBHmURpSHyQNic8a-t82Q. 

Thank you very much @sarah-clements for this great work 👍. ~~I have enough rights to land this patch. I'll handle this.~~ Edit: I forgot mergify handles this on this repo. I'm setting the correct flags then. 

I'm super excited to get Android-Components on shipit!
